web/__init__.py

```diff
@@ -32,60 +52,77 @@
 
 @app.before_first_request
 def startup():
-    app.svc.register("pppp", web.service.pppp.PPPPService())
-    app.svc.register("videoqueue", web.service.video.VideoQueue())
-    app.svc.register("mqttqueue", web.service.mqtt.MqttQueue())
-    app.svc.register("filetransfer", web.service.filetransfer.FileTransferService())
+    if app.config["login"]:
```

Comment:
I don't think this is a good way to do it.

These services are started on-demand, and should be registered as soon as possible, no matter if we are logged in or not.

Or course, we should be careful about not *requesting* them until we have login information ready.

Reply:
This should already be the case. The `app.svc.register("foo", web.service.x.y())` lines only register the service, it doesn't activate them, until they are requested.

If we don't register the services, and then (accidentally) hit an endpoint that requires the service, we are going to get an exception anyway, because the service is missing.

Maybe a better way to handle this in general, would be a decorator. Something like `@depends_on_service("foo")`, which could then make that service available as a parameter.

In such a case, we could properly handle missing services.

But until we have that, we should register the services as normal.

If you want to see some of the logic that starts services:

https://github.com/Ankermgmt/ankermake-m5-protocol/blob/8b2a8110c03e82e01256b9d1f47c9ca8eb6d9fa5/web/lib/service.py#L261

The `ServiceManager.get()` function bumps the reference count, then does:

```python
    ....
        if self.refs[name] == 1:
            svc.start()
```

So the service gets activated when it is requested, and not before.

Reply:
So if you don't register the services and attempt to hit the endpoint you get a 404, which is an error for the end user but not an Exception and doesn't cause the app to crash. If you register the service and there is no config you DO get an Exception and the app crashes as a result, causing a restart. You can't upload the config if the app is constantly restarting itself, this was my "simple" way of solving the problem.